File: frontend/src/api/sketchfab.ts
import { apiClient } from "./client.js";

export interface SketchfabStatus {
  enabled: boolean;
}

export interface SketchfabUploadRequest {
  name: string;
  description?: string;
  tags?: string[];
  is_private?: boolean;
  source_file?: string;
}

export interface SketchfabUploadResponse {
  job_id: string;
  status: string;
}

export interface SketchfabUploadStatus {
  job_id: string;
  status: string;
  sketchfab_uid?: string | null;
  sketchfab_url?: string | null;
  embed_url?: string | null;
  error_msg?: string | null;
}

export interface SketchfabImportRequest {
  url: string;
  name?: string | null;
}

export interface SketchfabImportResponse {
  asset_id: string;
}

export interface SketchfabModelItem {
  uid: string;
  name: string;
  url: string;
  thumbnail_url: string;
  vertex_count: number;
  face_count: number;
  is_downloadable: boolean;
  created_at: string;
}

export interface SketchfabMeModelsResponse {
  models: SketchfabModelItem[];
}

export async function getSketchfabStatus(): Promise<SketchfabStatus> {
  const { data } = await apiClient.get<SketchfabStatus>("/sketchfab/status");
  return data;
}

export async function uploadToSketchfab(
  assetId: string,
  body: SketchfabUploadRequest
): Promise<SketchfabUploadResponse> {
  const { data } = await apiClient.post<SketchfabUploadResponse>(
    `/assets/${assetId}/sketchfab/upload`,
    body
  );
  return data;
}

export async function getSketchfabUploadStatus(
  assetId: string
): Promise<SketchfabUploadStatus> {
  const { data } = await apiClient.get<SketchfabUploadStatus>(
    `/assets/${assetId}/sketchfab/status`
  );
  return data;
}

export async function importFromSketchfab(
  body: SketchfabImportRequest
): Promise<SketchfabImportResponse> {
  const { data } = await apiClient.post<SketchfabImportResponse>(
    "/assets/sketchfab/import",
    body
  );
  return data;
}

export async function getMySketchfabModels(): Promise<SketchfabMeModelsResponse> {
  const { data } = await apiClient.get<SketchfabMeModelsResponse>(
    "/sketchfab/me/models"
  );
  return data;
}

File: api/app/schemas/sketchfab.py
"""Pydantic-Schemas für Sketchfab-API."""

from pydantic import BaseModel, Field


class SketchfabUploadRequest(BaseModel):
    """Request für POST /assets/{asset_id}/sketchfab/upload."""

    name: str = Field(..., min_length=1, description="Modellname auf Sketchfab")
    description: str = Field(default="", description="Optionale Beschreibung")
    tags: list[str] = Field(default_factory=list, description="Tags, kommasepariert oder als Liste")
    is_private: bool = Field(default=False, description="Privat (True) oder öffentlich (False)")
    source_file: str = Field(default="mesh.glb", description="Welche GLB-Datei aus dem Asset-Ordner")


class SketchfabUploadResponse(BaseModel):
    """Response für POST /assets/{asset_id}/sketchfab/upload (202)."""

    job_id: str
    status: str = "pending"


class SketchfabUploadStatusResponse(BaseModel):
    """Response für GET /assets/{asset_id}/sketchfab/status."""

    job_id: str
    status: str
    sketchfab_uid: str | None = None
    sketchfab_url: str | None = None
    embed_url: str | None = None
    error_msg: str | None = None


class SketchfabImportRequest(BaseModel):
    """Request für POST /assets/sketchfab/import."""

    url: str = Field(..., description="Sketchfab-URL oder Model-UID")
    name: str | None = Field(default=None, description="Optionale Anzeige für das Asset")


class SketchfabImportResponse(BaseModel):
    """Response für POST /assets/sketchfab/import."""

    asset_id: str


class SketchfabModelItem(BaseModel):
    """Ein Modell aus GET /sketchfab/me/models."""

    uid: str
    name: str
    url: str
    thumbnail_url: str
    vertex_count: int
    face_count: int
    is_downloadable: bool
    created_at: str


class SketchfabMeModelsResponse(BaseModel):
    """Response für GET /sketchfab/me/models."""

    models: list[SketchfabModelItem]


class SketchfabStatusResponse(BaseModel):
    """Response für GET /sketchfab/status (Feature-Flag)."""

    enabled: bool
